libherokubuildpack/src/buildpack_output/util.rs

```diff
@@ -0,0 +1,223 @@
+use std::fmt::Debug;
+use std::io::Write;
+use std::sync::{Arc, Mutex};
+
+/// Iterator yielding every line in a string. Every line includes existing newline character(s).
+pub(crate) struct LineIterator<'a> {
+    input: &'a str,
+}
+
+impl<'a> LineIterator<'a> {
+    pub(crate) fn from(input: &'a str) -> LineIterator<'a> {
+        LineIterator { input }
+    }
+}
+
+impl<'a> Iterator for LineIterator<'a> {
+    type Item = &'a str;
+
+    #[inline]
+    fn next(&mut self) -> Option<&'a str> {
+        if self.input.is_empty() {
+            return None;
+        }
+
+        let newline_index = self.input.find('\n').map_or(self.input.len(), |i| i + 1);
+
+        let (line, rest) = self.input.split_at(newline_index);
+        self.input = rest;
+        Some(line)
+    }
+}
+
+pub(crate) fn prefix_lines<F: Fn(usize, &str) -> String>(contents: &str, f: F) -> String {
+    use std::fmt::Write;
+
+    let lines = LineIterator::from(contents).enumerate().fold(
+        String::new(),
+        |mut acc, (line_index, line)| {
+            let prefix = f(line_index, line);
+            let _ = write!(acc, "{prefix}{line}");
+            acc
+        },
+    );
+
+    if lines.is_empty() {
+        f(0, "")
+    } else {
+        lines
+    }
+}
+
+pub(crate) fn prefix_first_rest_lines(
+    first_prefix: &str,
+    rest_prefix: &str,
+    contents: &str,
+) -> String {
+    let first_prefix = String::from(first_prefix);
+    let rest_prefix = String::from(rest_prefix);
+
+    prefix_lines(contents, move |index, _| {
+        if index == 0 {
+            first_prefix.clone()
+        } else {
+            rest_prefix.clone()
+        }
+    })
+}
+
+#[derive(Debug)]
+pub(crate) struct LockedWriter<W> {
```

Comment:
[TLDR] I'm okay making it test only for now, and I don't want to delay this PR any more than needed, but I'm also not happy with the ergonomics of this interface for a gen 1 experience.

-----
Full: Originally I was wrapping the Write instance in this and returning it back to the user. Manuel changed that behavior so that `BuildpackOutput<state::Stream<W>>` is `Write` instead. It's currently not very ergonomic to use the streaming code with `std::process::Command` streaming because it requires two `Write` instances and we only give you one. So either we'll make the user have to implement something like this, or we will provide them with something like this, or we'll give them a different abstraction (Like a version of Command.stream that writes to both stdout and stderr with the same (locked) writer.

Another way to think about it: When I wrote all those doc tests, I did not show streaming a `Command` on purpose because it's currently very gnarly with this stripped down implementation and missing other pieces of support from the rest of libherokubuildpack.

Manuel expressed interest in this concept of a LockedWriter, presumably for the purpose of implementing a `Command` stream that accepts a single writer.

Reply:
Made it cfg(test)